File: C/Sorting/Tim_sort.c
#include <stdio.h>

void insertsort(int start, int end, int arr[])
{
    for (int i = start + 1; i <= end; i++)
    {
        int temp = arr[i];
        int j = i - 1;
        while (temp < arr[j] && j >= start)
        {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = temp;
    }
}
// Merge function merges the sorted runs
void merge(int arr[], int l, int m, int r)
{
 
    // Original array is broken in two
    // parts left and right array
    int len1 = m - l + 1, len2 = r - m;
    int left[len1], right[len2];
    for (int i = 0; i < len1; i++)
        left[i] = arr[l + i];
    for (int i = 0; i < len2; i++)
        right[i] = arr[m + 1 + i];
 
    int i = 0;
    int j = 0;
    int k = l;
 
    // After comparing, we
    // merge those two array
    // in larger sub array
    while (i < len1 && j < len2) {
        if (left[i] <= right[j]) {
            arr[k] = left[i];
            i++;
        }
        else {
            arr[k] = right[j];
            j++;
        }
        k++;
    }
 
    // Copy remaining elements of
    // left, if any
    while (i < len1) {
        arr[k] = left[i];
        k++;
        i++;
    }
 
    // Copy remaining element of
    // right, if any
    while (j < len2) {
        arr[k] = right[j];
        k++;
        j++;
    }
}

int main()
{
    // input code
    int n;
    printf("TIM SORT\nEnter array size : ");
    scanf("%d", &n);
    int ar[n];
    printf("\nEnter array : ");
    for (int i = 0; i < n; i++)
        scanf("%d", &ar[i]);

    // function call
    // insertsort(2,89,ar);
    //  display sorted array
    int runsize = 4;
    for (int start = 0; start < n; start = start + runsize)
    {
        int end;
        if (start + runsize - 1 < n - 1)
        {
            end = start + runsize - 1;
        }
        else
        {
            end = n - 1;
        }
      insertsort(start,end,ar);
    }for(int rs=runsize;rs<n;rs*=2){
    for(int left=0;left<n;left+=2*rs){
        int mid= left+rs-1;
        int right;
        if(left+2*rs-1<n-1){
            right=left+2*rs-1;
        }
        else{
            right=n-1;
        }if(mid<right){
            merge(ar,left,mid,right);
        }



    }
    }
    printf("\nSORTED ARRAY : ");
        for (int i = 0; i < n; i++)
            printf("%d ", ar[i]);

        return 0;
}

File: Python/Sorting/Tim_sort.py
# Python3 program to perform basic timSort
MIN_MERGE = 32
 
 
def calcMinRun(n):
    """Returns the minimum length of a
    run from 23 - 64 so that
    the len(array)/minrun is less than or
    equal to a power of 2.
 
    e.g. 1=>1, ..., 63=>63, 64=>32, 65=>33,
    ..., 127=>64, 128=>32, ...
    """
    right = 0
    while n >= MIN_MERGE:
        right |= n & 1
        n >>= 1
    return n + right
 
 
# This function sorts array from left index to
# to right index which is of size atmost RUN
def insertionSort(arr, left, right):
    for i in range(left + 1, right + 1):
        j = i
        while j > left and arr[j] < arr[j - 1]:
            arr[j], arr[j - 1] = arr[j - 1], arr[j]
            j -= 1
 
 
# Merge function merges the sorted runs
def merge(arr, left, mid, right):
 
    # original array is broken in two parts
    # left and right array
    len1, len2 = mid - left + 1, right - mid
    left, right = [], []
    for i in range(0, len1):
        left.append(arr[left + i])
    for i in range(0, len2):
        right.append(arr[mid + 1 + i])
 
    i, j, k = 0, 0, left
 
    # after comparing, we merge those two array
    # in larger sub array
    while i < len1 and j < len2:
        if left[i] <= right[j]:
            arr[k] = left[i]
            i += 1
 
        else:
            arr[k] = right[j]
            j += 1
 
        k += 1
 
    # Copy remaining elements of left, if any
    while i < len1:
        arr[k] = left[i]
        k += 1
        i += 1
 
    # Copy remaining element of right, if any
    while j < len2:
        arr[k] = right[j]
        k += 1
        j += 1
 
 
# Iterative Timsort function to sort the
# array[0...n-1] (similar to merge sort)
def timSort(arr):
    n = len(arr)
    minRun = calcMinRun(n)
 
    # Sort individual subarrays of size RUN
    for start in range(0, n, minRun):
        end = min(start + minRun - 1, n - 1)
        insertionSort(arr, start, end)
 
    # Start merging from size RUN (or 32). It will merge
    # to form size 64, then 128, 256 and so on ....
    size = minRun
    while size < n:
 
        # Pick starting point of left sub array. We
        # are going to merge arr[left..left+size-1]
        # and arr[left+size, left+2*size-1]
        # After every merge, we increase left by 2*size
        for left in range(0, n, 2 * size):
 
            # Find ending point of left sub array
            # mid+1 is starting point of right sub array
            mid = min(n - 1, left + size - 1)
            right = min((left + 2 * size - 1), (n - 1))
 
            # Merge sub array arr[left.....mid] &
            # arr[mid+1....right]
            if mid < right:
                merge(arr, left, mid, right)
 
        size = 2 * size


num = int(input("Enter the number of elements: "))
array = [int(x) for x in input("Enter list of numbers: ").split()]

timSort(array)

print(array)